Detectors/MUON/MCH/DevIO/Digits/digits-random-generator-workflow.cxx

```diff
@@ -0,0 +1,212 @@
+#include "Framework/ConfigParamSpec.h"
+#include "Framework/ControlService.h"
+#include "Framework/DataProcessorSpec.h"
+#include "Framework/DataSpecUtils.h"
+#include "Framework/Lifetime.h"
+#include "Framework/Logger.h"
+#include "Framework/Output.h"
+#include "Framework/Task.h"
+#include "Framework/WorkflowSpec.h"
+#include "MCHMappingInterface/Segmentation.h"
+#include "ProgOptions.h"
+#include <array>
+#include <chrono>
+#include <cstring>
+#include <fmt/format.h>
+#include <fstream>
+#include <iostream>
+#include <random>
+#include <stdexcept>
+#include <vector>
+
+/**
+ * This executable generate random MCH digits and their associated ROFRecords.
+ *
+ */
+
+using namespace o2::framework;
+
+using namespace o2::mch;
+
+namespace
+{
+std::array<int, 156> getDeIds()
+{
+  static std::array<int, 156> deids;
+  static bool first = true;
+  if (first) {
+    int i{0};
+    o2::mch::mapping::forEachDetectionElement([&](int deid) {
+      deids[i++] = deid;
+    });
+    first = false;
+  }
+  return deids;
+}
+
+std::array<int, 156> getNofPads()
+{
+  static bool first = true;
+  static std::array<int, 156> npads;
+  if (first) {
+    int i{0};
+    auto deids = getDeIds();
+    for (auto deid : deids) {
+      o2::mch::mapping::Segmentation seg(deid);
+      int16_t nofPads = static_cast<int16_t>(seg.nofPads());
+      npads[i++] = nofPads;
+    }
+    first = false;
+  }
+  return npads;
+}
+} // namespace
+
+constexpr const char* OPTNAME_NOF_ROFS_PER_TF = "nof-rofs-per-tf";
+constexpr const char* OPTNAME_OCCUPANCY = "occupancy";
+constexpr const char* OPTNAME_SEED = "seed";
+
+class DigitGenerator : public o2::mch::io::DigitIOBaseTask
+{
+ public:
+  void init(InitContext& ic)
+  {
+    io::DigitIOBaseTask::init(ic); // init common options
+    mOccupancy = ic.options().get<float>(OPTNAME_OCCUPANCY);
+    if (mOccupancy <= 0.0 || mOccupancy > 1.0) {
+      throw std::invalid_argument("occupancy must be between >0 and <=1");
+    }
+    mNofRofPerTimeFrame = ic.options().get<int>(OPTNAME_NOF_ROFS_PER_TF);
+    mSeed = ic.options().get<int>(OPTNAME_SEED);
+
+    if (!mSeed) {
+      std::random_device rd;
+      mSeed = rd();
+    }
+    mMersenneTwister.seed(mSeed);
+
+    LOGP(INFO,
+         "Will generate {:7.2f}% of pads in {:4d} ROFs "
+         "per timeframe, for {:4d} timeframes",
+         mOccupancy * 100.0, mNofRofPerTimeFrame, mMaxNofTimeFrames);
+  }
+
+  void run(ProcessingContext& pc)
+  {
+    if (mNofProcessedTFs >= mMaxNofTimeFrames) {
+      pc.services().get<ControlService>().endOfStream();
+      return;
+    }
+
+    if (shouldProcess()) {
+      incNofProcessedTFs();
+      std::vector<Digit> digits;
+      std::vector<ROFRecord> rofs;
+      int ndigits{0};
+
+      for (auto i = 0; i < mNofRofPerTimeFrame; i++) {
+        auto n = generateRandomDigits(mOccupancy, digits);
+        o2::InteractionRecord ir{mBC, mOrbit++};
+        rofs.emplace_back(ir, ndigits, n);
+        ndigits += n;
+        mOrbit++;
```

Comment:
mOrbit is incremented twice (here and 3 lines above)

Reply:
Good catch. Thanks.